node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -887,6 +939,22 @@ async fn process_incoming_peer_message<Context>(
 
 				peer_data.set_collating(collator_id, para_id);
 			} else {
+				if is_preconnect_allowed(ctx, runtime, &*state.view, para_id)
```

Comment:
`is_preconnect_allowed` is fairly heavy. I think a refactor so `allowed_preconnects: Set<ParaId>` are computed on `ActiveLeavesUpdate` and the `Declare` message just has a cheap set-membership check would be best.